File: src/core/check.ts
import {
  draftArtifactSchema,
  type CheckedArtifact,
  type DraftArtifact,
  type Finding,
} from './schemas.js';
import { compileBlueprint } from './blueprint/compile.js';
import { validateBlueprintRequest } from './blueprint/validate.js';
import { planIntentIssues } from './blueprint/plan-intent.js';
import { evaluateUnitDesign } from './blueprint/design-evaluation.js';
import { allLegalBuilds } from './mechanics/index.js';
import { candidateSchema } from './schemas.js';
import { freeze, verifyPrepared } from './prepare.js';
import { checkEvidence } from './check-evidence.js';
import { checkDependencies } from './check-dependencies.js';
import { checkProgression } from './check-progression.js';
import { findingSeverity, type ReportFinding } from './findings.js';

/** JSON object key order is not part of derived purchase evidence. Array order is. */
function orderedJson(value: unknown): string {
  return JSON.stringify(value, (_key, entry) =>
    entry && typeof entry === 'object' && !Array.isArray(entry)
      ? Object.fromEntries(
          Object.keys(entry)
            .sort()
            .map((key) => [key, entry[key]]),
        )
      : entry,
  );
}

/** Structural checks only. Natural-language semantics require separate review. */
export async function checkDraft(input: DraftArtifact): Promise<CheckedArtifact> {
  const draft = draftArtifactSchema.parse(input);
  await verifyPrepared(draft.prepared);
  const { candidate } = draft;
  const request = draft.prepared.request;
  const findings: Finding[] = [];
  const report: ReportFinding = (finding) => {
    findings.push({
      id: `deterministic.${findings.length + 1}`,
      method: 'deterministic',
      ...finding,
      severity: findingSeverity(finding.outcome),
      action: finding.action ?? null,
      evidence: finding.evidence ?? [],
    });
  };
  // Ordering is part of the retained artifact contract: IDs follow emitted findings.
  if (request.mechanicsDefinition) {
    const issues = candidate.blueprint
      ? validateBlueprintRequest(candidate.blueprint, request)
      : [
          {
            path: 'blueprint',
            message: 'A typed blueprint is required by this mechanics definition.',
          },
        ];
    if (
      candidate.blueprint &&
      issues.length === 0 &&
      JSON.stringify(candidateSchema.parse(compileBlueprint(candidate.blueprint, request))) !==
        JSON.stringify(candidate)
    )
      issues.push({
        path: 'candidate',
        message:
          'The readable candidate differs from its compiled blueprint. Recompile it instead of editing derived fields.',
      });
    if (
      candidate.blueprint &&
      issues.length === 0 &&
      draft.run.designEvaluation &&
      orderedJson(draft.run.designEvaluation) !==
        orderedJson(
          evaluateUnitDesign(
            candidate.blueprint,
            draft.run.designPlan,
            request.mechanicsDefinition,
          ),
        )
    )
      issues.push({
        path: 'run.designEvaluation',
        message:
          'The retained purchase evidence differs from the blueprint and plan. Recompute it instead of editing derived comparisons.',
      });
    if (issues.length)
      for (const issue of issues)
        report({
          category: 'conflict',
          outcome: 'fail',
          subject: issue.path,
          rule: 'typed-mechanics',
          message: issue.message,
          action: 'Correct the blueprint and compile again.',
        });
    else
      report({
        category: 'coverage',
        outcome: 'pass',
        subject: 'blueprint',
        rule: 'typed-mechanics',
        message: `All ${allLegalBuilds(request.mechanicsDefinition).length} legal builds resolve with valid stats, purchase gates, scoped boosts and matching compiled output. This does not simulate combat or certify balance.`,
      });
    if (issues.length === 0 && candidate.blueprint && draft.run.designPlan?.upgradeIntents)
      for (const issue of planIntentIssues(
        candidate.blueprint,
        draft.run.designPlan,
        request.mechanicsDefinition,
      ))
        report({
          category: 'conflict',
          outcome: 'fail',
          subject: issue.path,
          rule: 'planned-upgrade-intent',
          message: issue.message,
          action:
            'Implement the retained typed upgrade promise and compile again. This check does not assess prose, source interpretation or tactical value.',
        });
  }
  checkEvidence(candidate, request, report);
  checkDependencies(candidate, request, report);
  checkProgression(candidate, request.progression, report);
  if (!request.documents.some((document) => document.kind === 'rules')) {
    report({
      category: 'missing_specification',
      outcome: 'unresolved',
      subject: 'request.documents',
      rule: 'supplied-game-rules',
      message: 'No game rules document was supplied.',
      action: 'Supply governing rules before approving behavior.',
    });
  }
  report({
    category: 'scope',
    outcome: 'not_checked',
    subject: 'candidate',
    rule: 'validation-scope',
    message:
      (request.mechanicsDefinition
        ? 'Typed checks resolve every legal upgrade build and verify the compiled candidate. '
        : '') +
      'Deterministic checks cover references, assignments, dependencies and supplied ' +
      'progression constraints. They do not execute gameplay, establish semantic ' +
      'decision preservation, determine combat balance or grant ' +
      'acceptance.',
  });
  return freeze({
    schemaVersion: '1',
    kind: 'checked',
    draft,
    findings,
  });
}

File: internal/unit/check.go
package unit

import (
	"fmt"

	m "github.com/mardwerk/unit-generator/internal/mechanics"
	s "github.com/mardwerk/unit-generator/internal/schema"
)

// ParseDraft validates a draft artifact value.
func ParseDraft(value any) (Draft, error) {
	var d Draft
	return d, s.ParseInto(DraftSchema, value, &d)
}

// CheckDraft runs the deterministic checks. Findings keep their emitted order.
// Natural-language semantics require a separate review.
func CheckDraft(input Draft) (Checked, error) {
	draft, err := ParseDraft(s.FromGoValue(input))
	if err != nil {
		return Checked{}, err
	}
	if err := VerifyPrepared(draft.Prepared); err != nil {
		return Checked{}, err
	}
	candidate := draft.Candidate
	request := draft.Prepared.Request
	findings := []Finding{}
	report := func(f checkFinding) {
		evidence := f.Evidence
		if evidence == nil {
			evidence = []string{}
		}
		findings = append(findings, Finding{
			ID: fmt.Sprintf("deterministic.%d", len(findings)+1), Method: "deterministic",
			Category: f.Category, Severity: severity(f.Outcome), Outcome: f.Outcome, Subject: f.Subject,
			Rule: f.Rule, Message: f.Message, Evidence: evidence, Action: f.Action,
		})
	}
	if definition := request.MechanicsDefinition; definition != nil {
		var issues []m.Issue
		if candidate.Blueprint == nil {
			issues = []m.Issue{{Path: "blueprint", Message: "A typed blueprint is required by this mechanics definition."}}
		} else {
			issues = ValidateBlueprintRequest(*candidate.Blueprint, request)
		}
		if candidate.Blueprint != nil && len(issues) == 0 {
			compiled, err := CompileBlueprint(*candidate.Blueprint, request)
			if err != nil {
				return Checked{}, err
			}
			if s.Stringify(s.FromGoValue(compiled)) != s.Stringify(s.FromGoValue(candidate)) {
				issues = append(issues, m.Issue{Path: "candidate", Message: "The readable candidate differs from its compiled blueprint. Recompile it instead of editing derived fields."})
			}
		}
		if candidate.Blueprint != nil && len(issues) == 0 && draft.Run.DesignEvaluation != nil {
			evaluation, err := EvaluateUnitDesign(*candidate.Blueprint, draft.Run.DesignPlan, *definition)
			if err != nil {
				return Checked{}, err
			}
			if s.Canonical(draft.Run.DesignEvaluation) != s.Canonical(evaluation) {
				issues = append(issues, m.Issue{Path: "run.designEvaluation", Message: "The retained purchase evidence differs from the blueprint and plan. Recompute it instead of editing derived comparisons."})
			}
		}
		if len(issues) > 0 {
			for _, issue := range issues {
				report(checkFinding{Category: "conflict", Outcome: "fail", Subject: issue.Path, Rule: "typed-mechanics", Message: issue.Message, Action: act("Correct the blueprint and compile again.")})
			}
		} else {
			report(checkFinding{
				Category: "coverage", Outcome: "pass", Subject: "blueprint", Rule: "typed-mechanics",
				Message: fmt.Sprintf("All %d legal builds resolve with valid stats, purchase gates, scoped boosts and matching compiled output. This does not simulate combat or certify balance.", len(m.AllLegalBuilds(*definition))),
			})
		}
		if len(issues) == 0 && candidate.Blueprint != nil && draft.Run.DesignPlan != nil && draft.Run.DesignPlan.UpgradeIntents != nil {
			for _, issue := range PlanIntentIssues(*candidate.Blueprint, draft.Run.DesignPlan.UpgradeIntents, *definition) {
				report(checkFinding{
					Category: "conflict", Outcome: "fail", Subject: issue.Path, Rule: "planned-upgrade-intent", Message: issue.Message,
					Action: act("Implement the retained typed upgrade promise and compile again. This check does not assess prose, source interpretation or tactical value."),
				})
			}
		}
	}
	checkEvidence(candidate, request, report)
	checkDependencies(candidate, request, report)
	checkProgression(candidate, request.Progression, report)
	hasRules := false
	for _, d := range request.Documents {
		if d.Kind == "rules" {
			hasRules = true
		}
	}
	if !hasRules {
		report(checkFinding{
			Category: "missing_specification", Outcome: "unresolved", Subject: "request.documents", Rule: "supplied-game-rules",
			Message: "No game rules document was supplied.", Action: act("Supply governing rules before approving behavior."),
		})
	}
	scope := "Deterministic checks cover references, assignments, dependencies and supplied progression constraints. They do not execute gameplay, establish semantic decision preservation, determine combat balance or grant acceptance."
	if request.MechanicsDefinition != nil {
		scope = "Typed checks resolve every legal upgrade build and verify the compiled candidate. " + scope
	}
	report(checkFinding{Category: "scope", Outcome: "not_checked", Subject: "candidate", Rule: "validation-scope", Message: scope})
	return Checked{SchemaVersion: "1", Kind: "checked", Draft: draft, Findings: findings}, nil
}
